llama.cpp

```diff
@@ -447,726 +936,510 @@ struct llama_context {
-    std::vector<llama_load_tensor> tensors;
-    std::unordered_map<std::string, size_t> name_to_idx;
-};
+struct llama_model_loader {
+    int n_kv      = 0;
+    int n_tensors = 0;
+    int n_created = 0;
 
-enum llama_file_version {
-    LLAMA_FILE_VERSION_GGML,
-    LLAMA_FILE_VERSION_GGMF_V1, // added version field and scores in vocab
-    LLAMA_FILE_VERSION_GGJT_V1, // added padding
-    LLAMA_FILE_VERSION_GGJT_V2, // changed quantization format
-    LLAMA_FILE_VERSION_GGJT_V3, // changed Q4 and Q8 quantization format
-};
+    bool use_mmap = false;
 
-struct llama_file_loader {
     llama_file file;
     llama_file_version file_version;
-    llama_hparams hparams;
-    llama_vocab vocab;
-
-    llama_file_loader(const char * fname, llama_load_tensors_map & tensors_map)
-        : file(fname, "rb") {
-        LLAMA_LOG_INFO("llama.cpp: loading model from %s\n", fname);
-        read_magic();
-        read_hparams();
-        read_vocab();
-        read_tensor_metadata(tensors_map);
-    }
-    void read_magic() {
-        uint32_t magic = file.read_u32();
-
-        if (magic == LLAMA_FILE_MAGIC_GGML) {
-            file_version = LLAMA_FILE_VERSION_GGML;
-            return;
-        }
 
-        uint32_t version = file.read_u32();
+    std::unique_ptr<llama_mmap> mapping;
 
-        switch (magic) {
-            case LLAMA_FILE_MAGIC_GGMF:
-                switch (version) {
-                    case 1: file_version = LLAMA_FILE_VERSION_GGMF_V1; return;
-                }
-                break;
-            case LLAMA_FILE_MAGIC_GGJT:
-                switch (version) {
-                    case 1: file_version = LLAMA_FILE_VERSION_GGJT_V1; return;
-                    case 2: file_version = LLAMA_FILE_VERSION_GGJT_V2; return;
-                    case 3: file_version = LLAMA_FILE_VERSION_GGJT_V3; return;
-                }
-        }
+    struct gguf_context * ctx_gguf = NULL;
+    struct ggml_context * ctx_meta = NULL;
 
-        throw std::runtime_error(format("unknown (magic, version) combination: %08x, %08x; is this really a GGML file?",
-                     magic, version));
-    }
-    void read_hparams() {
-        hparams.n_vocab = file.read_u32();
-        hparams.n_embd  = file.read_u32();
-        hparams.n_mult  = file.read_u32();
-        hparams.n_head  = file.read_u32();
-        hparams.n_layer = file.read_u32();
-        hparams.n_rot   = file.read_u32();
-        hparams.ftype   = (enum llama_ftype) file.read_u32();
+    llama_model_loader(const std::string & fname, bool use_mmap) : file(fname.c_str(), "rb") {
+        struct gguf_init_params params = {
+            /*.no_alloc = */ true,
+            /*.ctx      = */ &ctx_meta,
+        };
 
-        // LLaMAv2
-        // TODO: read from header
-        hparams.n_head_kv = hparams.n_head;
-    }
-    void read_vocab() {
-        vocab.id_to_token.resize(hparams.n_vocab);
+        ctx_gguf = gguf_init_from_file(fname.c_str(), params);
```

Comment:
There's a missing check for `ctx_gguf == NULL` here, which is why "invalid magic number" leads to a segfault.